scripts/gb-model/compose_network.py

```diff
@@ -0,0 +1,311 @@
+        ppl,
+        str(hydro_profile),
+        str(hydro_capacities_path),
+        carriers,
+        **hydro_cfg,
+    )
+
+
+def add_gb_components(
+    n: pypsa.Network,
+    context: CompositionContext,
+) -> pypsa.Network:
+    if context.countries:
+        keep = n.buses.country.isin(context.countries)
+        drop = n.buses.index[~keep]
+        if len(drop) > 0:
+            logger.info("Removing %d buses outside target countries", len(drop))
+            n.mremove("Bus", drop)
+
+    meta = n.meta.setdefault("gb_model", {})
+    if context.countries:
+        meta["countries"] = list(context.countries)
+
+    return n
+
+
+def add_pypsaeur_components(
+    n: pypsa.Network,
+    config: dict[str, Any],
+    context: CompositionContext,
+    costs,
+) -> pypsa.Network:
+    if costs is not None:
+        add_electricity_grid_connection(n, costs)
+        n.meta.setdefault("gb_model", {})["costs_path"] = str(context.costs_path)
+
+    sanitize_locations(n)
+    try:
+        sanitize_carriers(n, config)
+    except KeyError as exc:  # pragma: no cover - tolerate partial configs
+        logger.debug("Skipping carrier sanitisation due to missing config: %s", exc)
+    return n
+
+
+def finalise_composed_network(
+    n: pypsa.Network,
+    context: CompositionContext,
+) -> pypsa.Network:
+    n.determine_network_topology()
+    meta = n.meta.setdefault("gb_model", {})
+    meta["resources_root"] = str(context.resources_root)
+    meta["composed"] = True
+    n.consistency_check()
+    return n
+
+
+def compose_network(
+    network_path: str,
+    output_path: str,
+    config: dict[str, Any],
+    inputs: Mapping[str, str],
+) -> None:
+    network = pypsa.Network(network_path)
+    context = create_context(network_path, config)
+    if "tech_costs" in inputs:
+        context.costs_path = Path(inputs["tech_costs"])
+    add_gb_components(network, context)
+
+    costs = None
+    costs_path = context.costs_path
+    if costs_path is not None and costs_path.exists():
+        weights = getattr(network.snapshot_weightings, "objective", None)
+        nyears = float(weights.sum()) / 8760.0 if weights is not None else 1.0
+        try:
+            costs = load_costs(
+                str(costs_path),
+                context.costs_config,
+                max_hours=context.max_hours,
+                nyears=nyears,
+            )
+        except Exception as exc:  # pragma: no cover - keep composing resilient
+            logger.warning(
+                "Failed to load cost data from %s: %s",
+                costs_path,
+                exc,
+            )
+    elif costs_path is not None:
+        logger.info(
+            "Cost data %s not found; skipping cost adjustments",
+            costs_path,
+        )
+
+    powerplants_path = Path(inputs["powerplants"]) if "powerplants" in inputs else None
+    hydro_capacities_path = (
+        Path(inputs["hydro_capacities"]) if "hydro_capacities" in inputs else None
+    )
```

Comment:
I'd prefer the script to fail on these not existing rather than (relatively silently) passing over them